src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -403,29 +403,34 @@ bool ObjectMonitor::enter(JavaThread* current) {
     }
 
     OSThreadContendState osts(current->osthread());
-    ThreadBlockInVM tbivm(current);
 
-    // TODO-FIXME: change the following for(;;) loop to straight-line code.
-    for (;;) {
-      current->set_suspend_equivalent();
-      // cleared by handle_special_suspend_equivalent_condition()
-      // or java_suspend_self()
+    assert(current->thread_state() == _thread_in_vm, "invariant");
 
+    current->frame_anchor()->make_walkable(current);
+    OrderAccess::storestore();
+    for (;;) {
+      current->set_thread_state(_thread_blocked);
       EnterI(current);
-
-      if (!current->handle_special_suspend_equivalent_condition()) break;
-
-      // We have acquired the contended monitor, but while we were
-      // waiting another thread suspended us. We don't want to enter
-      // the monitor while suspended because that would surprise the
-      // thread that suspended us.
-      //
-      _recursions = 0;
-      _succ = NULL;
-      exit(current, false /* not_suspended */);
-
-      current->java_suspend_self();
+      current->set_thread_state_fence(_thread_blocked_trans);
+      if (SafepointMechanism::should_process(current) &&
+        current->suspend_request_pending()) {
+        // We have acquired the contended monitor, but while we were
+        // waiting another thread suspended us. We don't want to enter
+        // the monitor while suspended because that would surprise the
+        // thread that suspended us.
+        _recursions = 0;
+        _succ = NULL;
+        exit(current, false /* not_suspended */);
+        SafepointMechanism::process_if_requested(current);
+        // Since we are going to _thread_blocked we skip setting _thread_in_vm here.
+      } else {
+        // Only exit path from for loop
+        SafepointMechanism::process_if_requested(current);
+        current->set_thread_state(_thread_in_vm);
+        break;
+      }
     }
+
     current->set_current_pending_monitor(NULL);
 
     // We cleared the pending monitor info since we've just gotten past
```

Comment:
Comment below needs updating since it mentions ThreadBlockInVM but we are not using it anymore.

Reply:
Fixed